repquota.1
\." $Id: repquota.1 7326 2008-05-02 16:04:05Z garlick $
.\"
.TH QUOTA 1 "Release 1.0" "" "QUOTA"
.SH NAME
repquota \- report file system quota information
.SH SYNOPSIS
.B repquota
.I "[--options] file-system"
.br
.SH DESCRIPTION
.B repquota
generates a report of quota limits and usage information for all users
of the specified file system.
.SH OPTIONS
.TP 
.I "-d,--dirscan"
Report on users who own top-level directories in the target file system.
Without this option, report on users in the password file.
.TP
.I "-b,--blocksize N"
Report disk usage in N-size units.  The suffixes `K', `M', or `G'
may be used for kilobytes, megabytes, or gigabytes, respectively.
Default: 1M.
.TP
.I "-m,--min-uid N"
Do not report on users whose UID is less than N.
.TP
.I "-M,--max-uid N"
Do not report on users whose UID is greater than N.
.TP
.I "-r,--reverse-sort"
Sort output in descending order.
.TP
.I "-s,--space-sort"
Sort output by space used (default is to sort on uid).
.TP
.I "-F,--files-sort"
Sort output by number of files used (default is to sort on uid).
.TP
.I "-u,--usage-only"
Only report usage information, not quota limits.
.TP
.I "-H,--suppress-heading"
Suppress printing of the initial heading in the report.
.TP
.I "-f configfile"
loads a configuration file other than the default (see FILES below).
.SH "FILES"
/etc/quota.conf
.SH "BUGS"
Any user can generate the quota report for NFS file systems.
Only root can generate the quota report for Lustre file systems.
.LP
Group quotas are not supported.
.SH "SEE ALSO"
quota.conf(5), quota(1)


NEWS
========================================================================
quota-1.9 Release Notes                                       2008-05-03
========================================================================

* Add support for Lustre quotas.

* Add 'repquota' command.

* General refactoring and cleanup.

========================================================================
quota-1.8 Release Notes                                       2008-04-18
========================================================================

* Fix bug where home directory of /g/g53/user matched mount point of
  /g/g5.  https://chaos.llnl.gov/bugzilla/show_bug.cgi?id=301

========================================================================
quota-1.7 Release Notes                                       2007-10-29
========================================================================

* Rename package to 'rquota' and have it obsolete the Red Hat quota 
  package.

========================================================================
quota-1.5 Release Notes                                       2006-10-06
========================================================================

* Added "-t sec" option which puts an upper limit on the amount of 
  time quota should be allowed to run, useful for login scripts.

* Added -l option to report only on the user's home directory.
  NOTE: this option determines the file system to query by comparing
  the user's home directory listed in the password file to the file 
  system labels in /etc/quota.conf.

* RPM packaging for AIX.


repquota.c
/*****************************************************************************\
 *  Copyright (C) 2001-2008 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Jim Garlick <garlick@llnl.gov>.
 *  UCRL-CODE-2003-005.
 *  
 *  This file is part of Quota, a remote quota program.
 *  For details, see <http://www.llnl.gov/linux/quota/>.
 *  
 *  Quota is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *  
 *  Quota is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *  
 *  You should have received a copy of the GNU General Public License along
 *  with Quota; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\*****************************************************************************/

#include <ctype.h>
#include <pwd.h>
#if HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#include <rpc/rpc.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/param.h>          /* MAXHOSTNAMELEN */
#include <signal.h>
#include <assert.h>
#include <dirent.h>
#include <libgen.h>
#include <sys/stat.h>

#include "getconf.h"
#include "getquota.h"
#include "util.h"
#include "list.h"

static void usage(void);
static void dirscan(confent_t *conf, List qlist, uid_t minuid, uid_t maxuid);
static void pwscan(confent_t *conf, List qlist, uid_t minuid, uid_t maxuid);

char *prog;

#define OPTIONS "m:M:b:dhHrsFf:u"
#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long(ac,av,opt,lopt,NULL)
static struct option longopts[] = {
    {"dirscan",          no_argument,        0, 'd'},
    {"blocksize",        required_argument,  0, 'b'},
    {"min-uid",          required_argument,  0, 'm'},
    {"max-uid",          required_argument,  0, 'M'},
    {"reverse",          no_argument,        0, 'r'},
    {"space-sort",       no_argument,        0, 's'},
    {"files-sort",       no_argument,        0, 'F'},
    {"usage-only",       no_argument,        0, 'u'},
    {"suppress-heading", no_argument,        0, 'H'},
    {"config",           required_argument,  0, 'f'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt(ac,av,opt)
#endif

int 
main(int argc, char *argv[])
{
    confent_t *conf;
    int c;
    int dopt = 0;
    unsigned long bsize = 1024*1024;
    uid_t minuid = 0;
    uid_t maxuid = 0;
    char *fsname;
    List qlist;
    int Fopt = 0;
    int ropt = 0;
    int sopt = 0;
    int Hopt = 0;
    int uopt = 0;

    prog = basename(argv[0]);
    while ((c = GETOPT(argc, argv, OPTIONS, longopts)) != EOF) {
        switch(c) {
            case 'd':   /* --dirscan */
                dopt++;
                break;
            case 'b':   /* --blocksize */
                if (parse_blocksize(optarg, &bsize)) {
                    fprintf(stderr, "%s: error parsing blocksize\n", prog); 
                    exit(1);
                }
                break;
            case 'm':   /* --min-uid */
                minuid = strtoul(optarg, NULL, 10);
                if (minuid == 0) {
                    fprintf(stderr, "%s: --min-uid must be >0\n", prog);
                    exit(1);
                }
                break;
            case 'M':   /* --max-uid */
                maxuid = strtoul(optarg, NULL, 10);
                if (maxuid == 0) {
                    fprintf(stderr, "%s: --max-uid must be >0\n", prog);
                    exit(1);
                }
                break;
            case 'r':   /* --reverse */
                ropt++;
                break;
            case 'F':   /* --files-sort */
                Fopt++;
                break;
            case 's':   /* --space-sort */
                sopt++;
                break;
            case 'u':   /* --usage-only */
                uopt++;
                break;
            case 'H':   /* --suppress-heading */
                Hopt++;
                break;
            case 'f':   /* --config */
                setconfent(optarg);
                break;
            case 'h':
            default:
                usage();
        }
    }
    if (Fopt && sopt) {
        fprintf(stderr, "%s: -f and -s are mutually exclusive\n", prog);
        exit(1);
    }
    if (optind < argc)
        fsname = argv[optind++];
    else
        usage();
    if (optind < argc)
        usage();
    if (!(conf = getconflabel(fsname))) {
        fprintf(stderr, "%s: %s: not found in quota.conf\n", prog, fsname);
        exit(1);
    }
    if (geteuid() != 0) {
        fprintf(stderr, "%s: must run with root privileges\n", prog);
        exit(1);
    }
        
    /* Scan.
     */
    qlist = list_create((ListDelF)quota_destroy);
    if (dopt) 
        dirscan(conf, qlist, minuid, maxuid);
    else
        pwscan(conf, qlist, minuid, maxuid);


    /* Sort.
     */
    if (ropt) {
        if (sopt)
            list_sort(qlist, (ListCmpF)quota_cmp_bytes_reverse);
        else if (Fopt)
            list_sort(qlist, (ListCmpF)quota_cmp_files_reverse);
        else
            list_sort(qlist, (ListCmpF)quota_cmp_uid_reverse);
    } else {
        if (sopt)
            list_sort(qlist, (ListCmpF)quota_cmp_bytes);
        else if (Fopt)
            list_sort(qlist, (ListCmpF)quota_cmp_files);
        else
            list_sort(qlist, (ListCmpF)quota_cmp_uid);
    }

    /* Report.
     */
    if (!Hopt) {
        char tmpstr[16];

        size2str(bsize, tmpstr, sizeof(tmpstr));
        printf("Quota report for %s (blocksize %s)\n", fsname, tmpstr);
    }
    if (uopt) {
        if (!Hopt)
            quota_report_heading_usageonly();
        list_for_each(qlist, (ListForF)quota_report_usageonly, &bsize);
    } else {
        if (!Hopt)
            quota_report_heading();
        list_for_each(qlist, (ListForF)quota_report, &bsize);
    }

    list_destroy(qlist);

    return 0;
}

static void
usage(void)
{
    fprintf(stderr, 
  "Usage: %s [--options] fs\n"
  "  -d,--dirscan           report on users who own top level dirs of fs\n"
  "  -b,--blocksize         report usage in blocksize units (default 1M)\n"
  "  -m,--min-uid           set minimum uid to include in report\n"
  "  -M,--max-uid           set maximum uid to include in report\n"
  "  -r,--reverse-sort      sort in reverse order\n"
  "  -s,--space-sort        sort on space used (default sort on uid)\n"
  "  -F,--files-sort        sort on files used (default sort on uid)\n"
  "  -u,--usage-only        only report usage, not quota limits\n"
  "  -H,--suppress-heading  suppress report heading\n"
  "  -f,--config            use a config file other than %s\n"
                , prog, _PATH_QUOTA_CONF);
    exit(1);
}

static void
dirscan(confent_t *cp, List qlist, uid_t minuid, uid_t maxuid)
{
    struct dirent *dp;
    DIR *dir;
    char fqp[MAXPATHLEN];
    struct stat sb;
    quota_t q;

    if (!(dir = opendir(cp->cf_rpath))) {
        fprintf(stderr, "%s: could not open %s\n", prog, cp->cf_rpath);
        exit(1);
    }
    while ((dp = readdir(dir))) {
        if (!strcmp(dp->d_name, ".") || !strcmp(dp->d_name, ".."))
            continue;
        snprintf(fqp, sizeof(fqp), "%s/%s", cp->cf_rpath, dp->d_name);
        if (stat(fqp, &sb) < 0)
            continue;
        if (minuid != 0 && sb.st_uid < minuid)
            continue;
        if (maxuid != 0 && sb.st_uid > maxuid)
            continue;
        if (list_find_first(qlist, (ListFindF)quota_match_uid, &sb.st_uid))
            continue;
        q = quota_create(cp->cf_label, cp->cf_rhost, cp->cf_rpath, 
                          cp->cf_thresh);
        if (quota_get(sb.st_uid, q)) {
            quota_destroy(q);
            continue; 
        }
        list_append(qlist, q);
    }
    if (closedir(dir) < 0)
        fprintf(stderr, "%s: closedir %s: %m\n", prog, cp->cf_rpath);
}

static void
pwscan(confent_t *cp, List qlist, uid_t minuid, uid_t maxuid)
{
    struct passwd *pw;
    quota_t q;

    while ((pw = getpwent()) != NULL) {
        if (minuid && pw->pw_uid < minuid)
            continue;
        if (maxuid != 0 && pw->pw_uid > maxuid)
            continue;
        if (list_find_first(qlist, (ListFindF)quota_match_uid, &pw->pw_uid))
            continue;
        q = quota_create(cp->cf_label, cp->cf_rhost, cp->cf_rpath, 
                          cp->cf_thresh);
        if (quota_get(pw->pw_uid, q)) {
            quota_destroy(q);
            continue;
        }
        list_append(qlist, q);
    }
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
